### PR TITLE
Replace `PaymentSheet.PaymentMethodType.dynamic` and other cases with `.stripe(STPPaymentMethod)` 

### PR DESCRIPTION
**Pull request by yuki-stripe**
Continues the `PaymentMethodType` refactor effort.  
> I'd like to stop using PaymentMethodType.dynamic and instead always use STPPaymentMethodType.  One reason is that doing so un-breaks the decoupled flow, where we pass an STPPaymentMethod object with type = unknown today for PMs that use PaymentMethodType.dynamic.  Another reason is that deleting PaymentMethodType.dynamic would also let us clean up the confusing STPPaymentMethodType <> PaymentMethodType conversion code.

Previous PR: https://github.com/stripe/stripe-ios/pull/2911 

This PR means we now require all new PaymentMethod types to be defined in STPPaymentMethodType.

## Motivation
I was able to remove or simplify lots of very tangled up and hard-to-reason-about code. 

As one example - `PaymentMethodType` has a `.bacsDebit` case, but you can also create `.dynamic("bacs_debit")`. There's no way to reason about the difference; different parts of the codebase treat these two cases slightly differently arbitrarily. The test `testBacsDDConfirmFlows` used `PaymentSheet.Type.dynamic("bacs_debit")` ([here](https://github.com/stripe/stripe-ios/blob/1f9baf3839380b9a5e3de45868dc74d00e6a4d33/StripePaymentSheet/StripePaymentSheetTests/PaymentSheet/PaymentSheetLPMConfirmFlowTests.swift#L173)), but the confirm code it wanted to test checks for `PaymentSheet.Type.bacsDebit` ([here](https://github.com/stripe/stripe-ios/blob/1f9baf3839380b9a5e3de45868dc74d00e6a4d33/StripePaymentSheet/StripePaymentSheet/Source/PaymentSheet/PaymentSheet%2BAPI.swift#L48)).  This means the test never actually entered the confirm code!

## Testing
- Existing automated tests
- Light manual testing of PaymentSheet, PaymentSheet.FlowController, CustomerSheet

## Changelog
Not user facing 🤞

### CODE REVIEW

File: StripePaymentSheet/StripePaymentSheetTests/PaymentSheet/PaymentSheetFormFactoryTest.swift [hunk 1]
```diff
@@ -40,47 +40,19 @@ class PaymentSheetFormFactoryTest: XCTestCase {
...
-        let params = form.updateParams(params: IntentConfirmParams(type: .dynamic("sepa_debit")))
+        let params = form.updateParams(params: IntentConfirmParams(type: .stripe(.SEPADebit)))
 
         XCTAssertEqual(params?.paymentMethodParams.billingDetails?.name, "Name")
         XCTAssertEqual(params?.paymentMethodParams.billingDetails?.email, "email@stripe.com")
         XCTAssertEqual(params?.paymentMethodParams.type, .SEPADebit)
-        XCTAssertEqual(params?.paymentMethodType, .dynamic("sepa_debit"))
-    }
-
-    func testSpecFromJSONProvider() {
```
[the remark this answers is not in the document]
Comment (yuki-stripe) [reply StripePaymentSheet/StripePaymentSheetTests/PaymentSheet/PaymentSheetFormFactoryTest.swift:58]: This test is redundant with the tests in FormSpecProviderTest and I removed specFromJSONProvider, which I think was the method this was testing.

File: StripePaymentSheet/StripePaymentSheetTests/PaymentSheet/PaymentSheetFormFactoryTest.swift [hunk 2]
```diff
@@ -89,10 +61,11 @@ class PaymentSheetFormFactoryTest: XCTestCase {
         let factory = PaymentSheetFormFactory(
             intent: .paymentIntent(STPFixtures.paymentIntent()),
             configuration: .paymentSheet(configuration),
-            paymentMethod: .dynamic("mock_payment_method")
+            paymentMethod: .stripe(.card)
```
[the remark this answers is not in the document]
Comment (yuki-stripe) [reply StripePaymentSheet/StripePaymentSheetTests/PaymentSheet/PaymentSheetFormFactoryTest.swift]: Thanks, I left a TODO for this but forgot to resolve.

File: StripePayments/StripePayments/Source/PaymentHandler/STPPaymentHandler.swift [hunk 3]
```diff
@@ -151,7 +151,7 @@ public class STPPaymentHandler: NSObject {
 
     private var formSpecPaymentHandler: FormSpecPaymentHandler?
 
-    internal var _redirectShim: ((URL, URL?, Bool) -> Void)?
+    public var _redirectShim: ((URL, URL?, Bool) -> Void)?
```
Comment (wooj-stripe) [inline StripePayments/StripePayments/Source/PaymentHandler/STPPaymentHandler.swift:154]: Are we sure we want to change this to public? This was largely here for testing.

Comment (yuki-stripe) [reply StripePayments/StripePayments/Source/PaymentHandler/STPPaymentHandler.swift:154]: Definitely don't want to make this public, thanks for catching!

**Review by wooj-stripe — APPROVED** [review]

**Pull request merged**